feat: run on desktop

RomanKovtyukh opened the pull request:
It includes sqlite configuration for windows, but only tested on Linux. Got an example from [here](https://github.com/alextekartik/flutter_app_example/blob/96cba15d72e82ad739323f489b261014e3854484/demo_sqflite/lib/main.dart#L15-L21) so decided not to remove the `Platform.isWindows` from there.

Before this change openreads was showing the following errors at the startup if you try running it on desktop:
```
flutter: [🌎 Easy Localization] [DEBUG] Start
[ERROR:flutter/runtime/dart_vm_initializer.cc(40)] Unhandled Exception: Bad state: databaseFactory not initialized
databaseFactory is only initialized when using sqflite. When using `sqflite_common_ffi`
You must call `databaseFactory = databaseFactoryFfi;` before using global openDatabase API

#0      databaseFactory.<anonymous closure> (package:sqflite_common/src/sqflite_database_factory.dart:29:7)
#1      databaseFactory (package:sqflite_common/src/sqflite_database_factory.dart:33:6)
#2      openDatabase (package:sqflite_common/sqflite.dart:83:10)
#3      DatabaseProvider.createDatabase (package:openreads/database/database_provider.dart:29:18)
<asynchronous suspension>
#4      DatabaseController.getAllBooks (package:openreads/database/database_controler.dart:30:16)
<asynchronous suspension>
#5      BookCubit._migrateCoversFromDatabaseToStorage (package:openreads/logic/cubit/book_cubit.dart:274:27)
<asynchronous suspension>
#6      BookCubit._initLoad (package:openreads/logic/cubit/book_cubit.dart:71:7)
<asynchronous suspension>

flutter: [🌎 Easy Localization] [DEBUG] Init state
```

#584

RomanKovtyukh committed:
feat: set up sqlite on linux and windows

It includes sql configuration for windows, but it's testted only on Linux.
Got an example from here: https://github.com/alextekartik/flutter_app_example/blob/96cba15d72e82ad739323f489b261014e3854484/demo_sqflite/lib/main.dart#L15-L21

Grandmother commented:
Checked the failures:
* will fix the code style in a separate PR
* the `intl` version was failing locally too until I built it with flutter 3.29.0. So we either need to migrate to both flutter 3.23 and newer intl version if we want to fix this or we need to fix the flutter version to 3.29.0 in CI.

Grandmother commented:
Ah, ok. Formatting is also changed because of the flutter version.

RomanKovtyukh committed:
feat: add linux platform to .metadata

RomanKovtyukh committed:
update flutter lints to 5.0.0

RomanKovtyukh committed:
Use blurhash_ffi to hash images on desktops

RomanKovtyukh committed:
Use showModalBottomSheet on all non-IOS systems

Grandmother commented:
Hi @mateusz-bak, could you please help reviewing this and couple more PRs? Or maybe there is someone else with write permissions who could help?

RomanKovtyukh committed:
Make AlertDialog cross-platform to support Linux

RomanKovtyukh committed:
ref: rename button builders to mean what they do

RomanKovtyukh committed:
Make reading tracker sreen work on linux

RomanKovtyukh committed:
Mark some functions as unimplemented on desktop platforms

RomanKovtyukh committed:
Use iOS implementation for import on desktops

RomanKovtyukh committed:
Don't show 'rate' links on desktops

mateusz-bak reviewed (changes requested):
Hi! Thanks for the work, looks like a good improvement.

However I don't see `linux` or  `windows` directories that I suppose were created after running  below command:
```
flutter create --platforms=linux,windows .
```

Also please rebase to fix conflicts on already done changes.

amoinier commented:
Hi! I would like give some help to merge this PR. Could you give me an update on the current status of the PR and any blockers you’re facing? Thanks in advance :)

RomanKovtyukh committed:
Add linux platform dir

RomanKovtyukh commented:
Hi, I've rebased the changes on top of the current HEAD and added the linux platform directory.

Regarding windows platform directory and build, I didn't add it because I haven't tested it there and won't  be able to.

One more thing that is missing yet is an automatic build configuration for linux. We won't be able to distribute the linux version without that.

armand-leguillou commented:
Hi! Has it evolved since last month?? (I can't wait to get OpenReads on my desktop bc i don't have a smartphone anymore)

RomanKovtyukh committed:
Add linux entry to test_build

RomanKovtyukh committed:
Fix camera icon for linux target

RomanKovtyukh commented:
BTW folks. While we're waiting for review you can try it out using the build artifacts: https://github.com/mateusz-bak/openreads/actions/runs/19399497883

To use it:
1. Download linux-appbundle.zip
2. unzip it somewhere on your linux machine
3. execute: `/path/to/openreads`

I did fix everything platform-specific I could fine, but I might've missed something. So let me know if it's crashing or silently ignoring you.

CC: @armand-leguillou, @amoinier, @ironhak, @GLLM

mateusz-bak commented:
Great work @RomanKovtyukh I'm starting the review, I'll try to finish it by tomorrow.

mbeko commented on lib/ui/settings_screen/settings_backup_screen.dart:
```diff
@@ -118,14 +129,21 @@ class _SettingsBackupScreenState extends State<SettingsBackupScreen> {
   _startCreatingCloudBackup(context) async {
+    if (!(Platform.isAndroid || Platform.isIOS)) {
+      BackupGeneral.showInfoSnackbar(
+        'Cloud backup is not implemented for ${Platform.operatingSystem} yet.',
```
This shows "Cloud backup is not implemented for linux yet."
I think it should be "Linux"

RomanKovtyukh replied:
Moved to translation file and to avoid concatenating strings it's not mentioning platform any more.

RomanKovtyukh commented:
Thanks for the review, @mbeko. Really good findings you have there.

> From the welcome pages, I tried to import a backup that I created on Android, and it said that it's not valid. When I tried it from the settings menu later, it worked

I'm kind of able to reproduce this or similar issue sometimes: openreads is saying "the file is not a valid opereads backup", but then it's working perfectly fine on subsequent runs. Debugging showed that the file is really the right backup, so we must've returned `nil` instead of a version from the [`_checkInfoFileVersion`](https://github.com/mateusz-bak/openreads/blob/e54ad3771ea5bb1852ec34d600e73cd2a73d50ae/lib/core/helpers/backup/backup_import.dart#L458) and that looks like some race condition to me.

> Nothing happens when clicking on the button to scan the ISBN barcode. There should be at least a message saying that this isn't available on Linux. If the functionality actually is available, but there's no active camera, the message should say that a
camera is needed and must be enabled

Sure, need to fix it before releasing.

> When clicking on the options to create a local backup or to export to CSV, it says that these aren't implemented on Linux yet, but the icons turn into progress animations that don't stop. Then I noticed the laptop's fans started spinning, and the CPU
usage went from around 0% to 60-80%. This doesn't happen for the cloud backup option

Yeah, need to fix this. I've seen it, but didn't pay attention since I didn't wait long enough to notice the CPU consumption.

> It's not clear to me why the local backup or CSV export wouldn't work on desktop because that doesn't seem platform-specific, and I think these would be quite important to have

Agree

> It seems files like the book database and the cover images are put directly in the home folder. This is usually frowned upon, and the recommended practice is to follow the XDG base directory specification. These files would go to $XDG_DATA_HOME (~/.
local/share by default) in my opinion. I don't know if there are more files that are created by the app. If so, they should be put in an appropriate folder, too

Good point. I need to check since I've seen files in `~/.local/` too. Would be good to have that as part settings eventually so that we can share the files with [syncthing on mobile platforms](https://github.com/mateusz-bak/openreads/discussions/714).

> I didn't find the settings button at first because it's covered by the "Debug" indicator. I guess this won't be a problem in the final release, but when we disable the debug mode, we should check that the button is actually displayed

Yeah, had the same trouble. On the release builds it's working well -- just no "Debug" indicator.

Still need to check the rest of the issues you mentioned. Thanks again for the review.

mbeko commented:
I imagine this pull request will take some days before it can be merged. So how about submitting a patch, and if it's not accepted by the time we're ready to merge, we'll capitalise ourselves? I'd definitely not hardcode it

RomanKovtyukh committed:
bugfix: backup extraction race condition

Make unzipping the archive synchronous to avoid reading
files from backup before they got extracted.

In the backup v5 the `info.txt` file is getting extracted
first to avoid extracting the whole archive before it's
confirmed to be a valid backup.

RomanKovtyukh commented:
Made the extraction fully synchronous everywhere. In the `_restoreBackupVersion5` extracted `info.txt` first to confirm the validity of the backup before extracting the whole archive.

Made it so because even after extracting `into.txt` first I noticed that some of the images were used by the [backup restoration](https://github.com/mateusz-bak/openreads/blob/e54ad3771ea5bb1852ec34d600e73cd2a73d50ae/lib/core/helpers/backup/backup_import.dart#L197) before they got extracted.

RomanKovtyukh commented:
> When the backup import didn't work on the welcome pages, I tried to import the CSV, but when I clicked on the button, no file selection dialogue opened. Also from the settings menu, it doesn't work. Actually, none of the three CSV import options shows a dialogue

Added the same "not implemented" note to all.

RomanKovtyukh committed:
Use XDG_DATA_HOME as default on Linux platform

mateusz-bak commented on lib/ui/welcome_screen/welcome_screen.dart:
```diff
@@ -102,6 +106,10 @@ class _WelcomeScreenState extends State<WelcomeScreen> {
       await CSVImportGoodreads.importCSV(context);
+    } else {
+      BackupGeneral.showInfoSnackbar(
+        'Goodreads CSV import is not implemented for ${Platform.operatingSystem} yet.',
```
Please move string to translations file

mateusz-bak commented on lib/ui/welcome_screen/welcome_screen.dart:
```diff
@@ -102,6 +106,10 @@ class _WelcomeScreenState extends State<WelcomeScreen> {
       await CSVImportGoodreads.importCSV(context);
+    } else {
+      BackupGeneral.showInfoSnackbar(
+        'Goodreads CSV import is not implemented for ${Platform.operatingSystem} yet.',
```
Please move string to translations file

RomanKovtyukh replied:
Done

mateusz-bak commented on lib/ui/welcome_screen/welcome_screen.dart:
```diff
@@ -120,6 +128,10 @@ class _WelcomeScreenState extends State<WelcomeScreen> {
       await CSVImportBookwyrm.importCSV(context);
+    } else {
+      BackupGeneral.showInfoSnackbar(
+        'BookWyrm CSV import is not implemented for ${Platform.operatingSystem} yet.',
```
Please move string to translations file

RomanKovtyukh replied:
Done

mateusz-bak commented on lib/database/database_provider.dart:
```diff
@@ -19,7 +20,10 @@ class DatabaseProvider {
 
   Future<Database> createDatabase() async {
-    Directory docDirectory = await getApplicationDocumentsDirectory();
+    Directory docDirectory = Directory('${dataHome.path}/openreads');
```
As package xdg_directories supports only Linux maybe we should set docDirectory as:
```
Directory docDirectory;

if (Platform.isLinux) {
  docDirectory = Directory('${dataHome.path}/openreads');
} else {
  docDirectory = await getApplicationDocumentsDirectory();
}
```

RomanKovtyukh replied:
Agree. Done.

mateusz-bak commented on lib/main.dart:
```diff
@@ -49,16 +51,25 @@ void main() async {
+
   _setAndroidConfig();
 
+  appDocumentsDirectory = Directory('${dataHome.path}/openreads');
```
As package xdg_directories supports only Linux maybe we should set appDocumentsDirectory as:

```
appDocumentsDirectory = Directory('${dataHome.path}/openreads');
if (Platform.isLinux) {
  appDocumentsDirectory = Directory('${dataHome.path}/openreads');
} else {
  appDocumentsDirectory = await getApplicationDocumentsDirectory();
}
```

RomanKovtyukh replied:
Sorry, messed up here. Not sure why I made the `Directory('${dataHome.path}/openreads')` as default for non-ios/android.

Regarding what you've proposed, totally agree, but I'd write it as follows:
```dart
  appDocumentsDirectory = await getApplicationDocumentsDirectory();
  if (Platform.isLinux) {
    appDocumentsDirectory = Directory('${dataHome.path}/openreads');
```

mateusz-bak commented on lib/ui/settings_screen/settings_backup_screen.dart:
```diff
@@ -86,6 +95,10 @@ class _SettingsBackupScreenState extends State<SettingsBackupScreen> {
       await CSVImportGoodreads.importCSV(context);
+    } else {
+      BackupGeneral.showInfoSnackbar(
+        'Goodreads CSV import is not implemented for ${Platform.operatingSystem} yet.',
```
Please move string to translations file.

RomanKovtyukh replied:
Done

mateusz-bak commented on lib/ui/settings_screen/settings_backup_screen.dart:
```diff
@@ -103,6 +116,10 @@ class _SettingsBackupScreenState extends State<SettingsBackupScreen> {
       await CSVImportBookwyrm.importCSV(context);
+    } else {
+      BackupGeneral.showInfoSnackbar(
+        'BookWyrm CSV import is not implemented for ${Platform.operatingSystem} yet.',
```
Please move string to translations file.

RomanKovtyukh replied:
Done

mateusz-bak commented on lib/ui/settings_screen/settings_backup_screen.dart:
```diff
@@ -120,12 +137,23 @@ class _SettingsBackupScreenState extends State<SettingsBackupScreen> {
       await CSVImportOpenreads.importCSV(context);
+    } else {
+      BackupGeneral.showInfoSnackbar(
+        'Openreads CSV import is not implemented for ${Platform.operatingSystem} yet.',
```
Please move string to translations file.

RomanKovtyukh replied:
Done

mateusz-bak commented on lib/ui/add_book_screen/add_book_screen.dart:
```diff
@@ -593,6 +594,8 @@ class _AddBookScreenState extends State<AddBookScreen> {
                         borderRadius: BorderRadius.circular(cornerRadius),
                       ),
                       onTap: () async {
+                        if (!Platform.isAndroid && !Platform.isIOS) return;
```
Instead of doing that please hide the `InkWell` widget on platforms other than Android and iOS until this is implemented.

RomanKovtyukh replied:
Done. Used `Visibility` widget.

mateusz-bak reviewed (changes requested):
@RomanKovtyukh awesome work! I'm very sorry about the delay (I'm only starting to acclimate to new reality of life with a newborn).

I left some comments about the code - couple of simple fixes that I would like to see. I'll go through comments on the PR and will gather my thought about the Linux specific stuff and UI/UX. 

But I think after the requested changes we can merge this PR and have a list of things to improve in the Linux version.

Again great work, I really appreciate it!

mateusz-bak commented:
> Ah, I see. I'm unfortunately stuck with 2.9.0 on Android so I can't compare, but then it's likely that it's not an issue specific to the desktop version. Thanks!

@mbeko this is exactly how it is implemented right now. From Book Screen you can click on tags to show list of books that have the tag assigned. From Home Screen the tags are not clickable. If you think it should be added, please raise as a feature request in the Discussions :blue_heart:

RomanKovtyukh commented on lib/core/helpers/helpers.dart:
```diff
@@ -14,11 +15,25 @@ import 'package:openreads/main.dart';
+      Constants.blurHashY,
+    );
+  } else {
+    throw Exception('Unsupported platform for generating BlurHash');
```
Done

mateusz-bak commented:
@RomanKovtyukh we are ready to merge this PR when you implement small changes I requested. I'm holding all other PRs so we don't need to rebase this one again.

RomanKovtyukh commented:
Hi. Let me address the issues this week.

RomanKovtyukh commented on lib/ui/settings_screen/settings_backup_screen.dart:
```diff
@@ -52,6 +53,10 @@ class _SettingsBackupScreenState extends State<SettingsBackupScreen> {
       await BackupExport.createLocalBackup(context);
+    } else {
+      BackupGeneral.showInfoSnackbar(
+        'Backup export is not implemented for ${Platform.operatingSystem} yet.',
```
Done

RomanKovtyukh commented on lib/ui/settings_screen/settings_backup_screen.dart:
```diff
@@ -69,6 +74,10 @@ class _SettingsBackupScreenState extends State<SettingsBackupScreen> {
       await CSVExport.exportCSV();
+    } else {
+      BackupGeneral.showInfoSnackbar(
+        'CSV export is not implemented for ${Platform.operatingSystem} yet.',
```
Done

RomanKovtyukh commented on lib/ui/welcome_screen/welcome_screen.dart:
```diff
@@ -84,6 +84,10 @@ class _WelcomeScreenState extends State<WelcomeScreen> {
       await CSVImportOpenreads.importCSV(context);
+    } else {
+      BackupGeneral.showInfoSnackbar(
+        'Openreads CSV import is not implemented for ${Platform.operatingSystem} yet.',
```
Done

RomanKovtyukh commented on lib/ui/search_ol_screen/search_ol_screen.dart:
```diff
@@ -244,6 +244,8 @@ class _SearchOLScreenState extends State<SearchOLScreen>
   }
 
   void _startScanner() async {
+    if (!Platform.isAndroid && !Platform.isIOS) return;
```
Done

mateusz-bak commented:
Hello @RomanKovtyukh  thank you for the changes! I think hiding will be better but I'm afraid we will loose the visibility and never fix them. If you hide the elements can you open a new issue to implement them properly on desktop? I think one ticket with all the features hidden will be ok.

RomanKovtyukh commented:
And agree about visibility too. Will create a ticket with everything we decide to hide here.

mateusz-bak commented:
Agree! Greyed out buttons would be best! We can implement them after we merge this PR.

RomanKovtyukh commented:
Ok, made inactive buttons transparent and with same info bar showing a warning message (translated). It's quite ugly in code, but hopefully will be fixed quickly after merging this PR because most of the things are about working with CSVs which I expect to be fixed all together.

mateusz-bak commented on assets/translations/en-US.json:
```diff
@@ -289,5 +289,14 @@
-}
\ No newline at end of file
+    "show_title_over_cover": "Show title over cover",
+    "unsupported_platform_blur_hash_error": "Unsupported platform for generating BlurHash",
```
Please combine all unsuported... strings to one:
```
This action is not supported on current platform
```
This will make life easier for the translator especially that we are going to support the features soon.

mateusz-bak commented on lib/core/helpers/helpers.dart:
```diff
@@ -14,17 +15,36 @@ import 'package:openreads/main.dart';
+      Constants.blurHashX,
+      Constants.blurHashY,
+    );
+  } else {
```
I don't think we need to have the else statement. When we have new platforms we will review the platform specific implementations so no point in keeping the exception placeholders..

mateusz-bak reviewed (changes requested):
Great work, I added small comments. Please fix and also rebase

RomanKovtyukh commented:
Addressed the last comments, @mateusz-bak.

RomanKovtyukh commented:
I'm checking the CSV export and not sure if I understand right what [this](https://github.com/mateusz-bak/openreads/blob/master/lib/core/helpers/backup/csv_export.dart#L66-L68) is doing on iOS:

```
        await SharePlus.instance.share(ShareParams(
          files: [XFile(path)],
        ));
```
and how should I work it around on Linux (SharePlus doesn't implement fire share on Linux).

Seem like it's opening sharing menu after saving CSV file, but I can't check it.

Do we need to do that on Android too? Or it's done on iOS because it's not straightforward there where the file was saved and users in most cases would have to choose another path?

Other than that it's working ok on Linux, so I event can add to this PR or open another one right after this.
